app/models/message_notifier.rb

```diff
@@ -0,0 +1,90 @@
+class MessageNotifier
+  def initialize(user, ability, facility)
+    @user = user
+    @ability = ability
+    @facility = facility
+  end
+
+  def message_count
+    notifications.count +
+    order_details_in_dispute.count +
+    problem_order_details.count +
+    problem_reservation_order_details.count +
+    training_requests.count
+  end
+
+  def messages?
+    notifications? ||
```

Comment:
What if this method was just `message_count > 0` and `message_count` was memoized? It looks like you handle any of the permissions there as well.

It may be premature optimization, but what if you memoized each individual count? I think that would minimize the number of queries run (one `count(*)` for each notification type).